File: Web/src/domain/instrument/hole.ts
import type { PositionOnHarmonica } from '../playing/positionOnHarmonica.js'
import { isOnTheHarmonica } from '../playing/positionOnHarmonica.js'
import type { MouthWidth } from '../playing/mouthWidth.js'

export type Hole = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10

export const holes: readonly Hole[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

const edgeTolerance = 1e-9

export function holeAt(position: PositionOnHarmonica): Hole | null {
    if (!isOnTheHarmonica(position)) return null

    return holes[nearestIndex(position.fractionFromLeftEdge)] ?? null
}

export function holesCovered(position: PositionOnHarmonica, width: MouthWidth): Hole[] {
    if (!isOnTheHarmonica(position)) return []

    if (width === 1) {
        const single = holeAt(position)
        return single === null ? [] : [single]
    }

    const centre = position.fractionFromLeftEdge * holes.length
    const leftEdge = centre - width / 2
    const rightEdge = centre + width / 2
    return holes.filter((_, index) => isWithin(index + 0.5, leftEdge, rightEdge))
}

function isWithin(centre: number, leftEdge: number, rightEdge: number): boolean {
    return centre >= leftEdge - edgeTolerance && centre < rightEdge - edgeTolerance
}

function nearestIndex(fraction: number): number {
    const withinTheStrip = Math.min(1, Math.max(0, fraction))
    return Math.min(holes.length - 1, Math.floor(withinTheStrip * holes.length))
}

File: Core/Sources/HarmonicaCore/Entities/Instrument/Hole.swift
public enum Hole: Int, CaseIterable, Sendable {
    case one = 1
    case two
    case three
    case four
    case five
    case six
    case seven
    case eight
    case nine
    case ten

    private static let edgeTolerance = 1e-9

    // MARK: - Public

    public init?(at position: PositionOnHarmonica) {
        guard position.isOnTheHarmonica else { return nil }

        self = Hole.allCases[Hole.nearestIndex(atFraction: position.fractionFromLeftEdge)]
    }

    public static func allCovered(by position: PositionOnHarmonica, measuring mouth: MouthMeasure) -> [Hole] {
        guard position.isOnTheHarmonica else { return [] }

        switch mouth {
        case .theContactItself: return coveredByTheContact(at: position)
        case .holesWide(let width): return covered(by: width, centredAt: position)
        }
    }

    public var number: Int {
        rawValue
    }

    // MARK: - Private

    private static func coveredByTheContact(at position: PositionOnHarmonica) -> [Hole] {
        let leftmost = nearestIndex(atFraction: position.fractionFromLeftEdge - position.fractionCoveredEitherSide)
        let rightmost = nearestIndex(atFraction: position.fractionFromLeftEdge + position.fractionCoveredEitherSide)
        return (leftmost...rightmost).map { allCases[$0] }
    }

    private static func covered(by width: MouthWidth, centredAt position: PositionOnHarmonica) -> [Hole] {
        guard width != .oneHole else { return Hole(at: position).map { [$0] } ?? [] }

        let centre = position.fractionFromLeftEdge * Double(allCases.count)
        let span = Double(width.holes) / 2
        return allCases.indices
            .filter { isCovered(centreOfTheHole: Double($0) + 0.5, from: centre - span, to: centre + span) }
            .map { allCases[$0] }
    }

    private static func isCovered(centreOfTheHole centre: Double, from leftEdge: Double, to rightEdge: Double) -> Bool {
        centre >= leftEdge - edgeTolerance && centre < rightEdge - edgeTolerance
    }

    private static func nearestIndex(atFraction fraction: Double) -> Int {
        let withinTheStrip = min(1, max(0, fraction))
        return min(allCases.count - 1, Int(withinTheStrip * Double(allCases.count)))
    }
}
